### PR TITLE
Port FIRQuery to C++

### PR DESCRIPTION
**Pull request by wilhuff**
Note that this adds a `Direction` enum, to bring this implementation slightly closer to the Android one. This will be used more when we port `FSTQuery`.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/api/query_core.h [hunk 1]
```diff
@@ -0,0 +1,192 @@
+/*
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.h]: Question: just curious, why is this file called `query_core` and not just `query`?

File: Firestore/Source/Core/FSTFirestoreClient.mm [hunk 2]
```diff
@@ -393,13 +393,11 @@ - (void)getDocumentsFromLocalCache:(FIRQuery *)query
     ViewSnapshot snapshot = std::move(viewChange.snapshot).value();
     SnapshotMetadata metadata(snapshot.has_pending_writes(), snapshot.from_cache());
 
-    FIRQuerySnapshot *result = [[FIRQuerySnapshot alloc] initWithFirestore:query.firestore.wrapped
-                                                             originalQuery:query.query
-                                                                  snapshot:std::move(snapshot)
-                                                                  metadata:std::move(metadata)];
+    api::QuerySnapshot result(query.firestore(), query.query(), std::move(snapshot),
+                              std::move(metadata));
```
Comment (var-const) [inline Firestore/Source/Core/FSTFirestoreClient.mm]: Nit: I think `metadata` is a POD, in which case I'd rather not `move` it (I think it applies in more than one spot in this PR).

File: Firestore/core/src/firebase/firestore/api/query_core.mm [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    std::promise<ListenerRegistration> registration_promise_;
+  };
+
+  auto listener = absl::make_unique<ListenOnce>(source, std::move(callback));
+  auto listener_unowned = listener.get();
+
+  ListenerRegistration registration =
+      AddSnapshotListener(std::move(options), std::move(listener));
+
+  listener_unowned->Resolve(std::move(registration));
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.mm]: Not for this PR: this general pattern has a lot of resemblance to `api::DocumentReference::GetDocument`. Perhaps there is a way to abstract some part of it away.

File: Firestore/core/src/firebase/firestore/api/query_core.mm [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+  return ListenerRegistration(firestore_->client(), std::move(async_listener),
+                              std::move(query_listener));
+}
+
+Query Query::Filter(FieldPath field_path,
+                    Filter::Operator op,
+                    FSTFieldValue* field_value,
+                    const std::function<std::string()>& type_describer) const {
+  if (field_path.IsKeyFieldPath()) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.mm]: Not for this PR: this function is somewhat hard to follow (3-level-nested `if` statements in particular).

File: Firestore/core/src/firebase/firestore/api/query_core.mm [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        ThrowInvalidArgument(
+            "Invalid query. When querying a collection by document ID you must "
+            "provide a plain document ID, but '%s' contains a '/' character.",
+            document_key);
+      }
+      ResourcePath path =
+          query().path.Append(ResourcePath::FromString(document_key));
+      if (!DocumentKey::IsDocumentKey(path)) {
+        ThrowInvalidArgument(
+            "Invalid query. When querying a collection group by document ID, "
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.mm]: Question: why is it certain that this is a collection group query?

File: Firestore/core/src/firebase/firestore/api/query_core.mm [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+Query Query::EndAt(FSTBound* bound) const {
+  return Wrap([query() queryByAddingEndAt:bound]);
+}
+
+void Query::ValidateNewRelationFilter(FSTRelationFilter* filter) const {
+  if ([filter isInequality]) {
+    const FieldPath* existingField = [query_ inequalityFilterField];
+    if (existingField && *existingField != filter.field) {
+      ThrowInvalidArgument(
+          "Invalid Query. All where filters with an inequality (lessThan, "
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.mm]: Optional: is it possible to emphasize `where` so that it's clear that it's a "keyword"? E.g. by putting it into single quotes, etc.

File: Firestore/core/src/firebase/firestore/api/query_core.mm [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+      ValidateOrderByField(fieldPath, *inequalityField);
+    }
+  }
+}
+
+void Query::ValidateOrderByField(const FieldPath& orderByField,
+                                 const FieldPath& inequalityField) const {
+  if (orderByField != inequalityField) {
+    ThrowInvalidArgument(
+        "Invalid query. You have a where filter with an inequality (lessThan, "
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/api/query_core.mm]: Same nit about `'where'`.

File: Firestore/core/src/firebase/firestore/api/query_core.mm (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/api/query_core.mm]: Agreed in principle, but things are just different enough that this is not easy.

File: Firestore/core/src/firebase/firestore/api/query_core.mm (hunk 6 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/api/query_core.mm]: Done.

File: Firestore/core/src/firebase/firestore/core/direction.h [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+#include "absl/base/attributes.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+/** Interface used for all query orderings. All Directions are immutable. */
+class Direction {
+ public:
+  ABSL_CONST_INIT static const Direction Ascending;
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/core/direction.h]: I haven't been able to make that work. The compiler complains that it can't make incomplete types `constexpr` (and indicates that the type is incomplete until the closing `};`).

If I were willing to put these outside the class this could work, but most of the point of this is to make them scoped so I think this is likely the best form.

File: Firestore/core/src/firebase/firestore/core/direction.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    return comparison_modifier_;
+  }
+
+ private:
+  enum {
+    AscendingModifier = 1,
+    DescendingModifier = -1,
+  };
+
+  constexpr explicit Direction(int comparison_modifier) noexcept
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/core/direction.h]: None really. Removed.

File: Firestore/core/src/firebase/firestore/api/query_core.mm (hunk 5 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/api/query_core.mm]: I'm not entirely sure. I've double-checked that the logic is unmodified from the source. @mikelehen, could you comment on why this is certainly a collection group query at this point? That doesn't seem to be supported by the ifs above.

File: Firestore/Source/Core/FSTFirestoreClient.mm (hunk 2 above)
Comment (var-const) [reply Firestore/Source/Core/FSTFirestoreClient.mm]: I can see the argument going both ways, so ultimately I'll leave this to your judgment. One could argue that `move`ing PODs reduces clarity by making it appear that the class in question is more complicated than it actually is. `clang-tidy` [flags this](https://clang.llvm.org/extra/clang-tidy/checks/performance-move-const-arg.html), presumably on these grounds. On the other hand, it's a good point that `move`ing such types is more future-proof. Feel free to leave as is.

File: Firestore/core/src/firebase/firestore/api/query_core.mm (hunk 5 above)
Comment (mikelehen) [reply Firestore/core/src/firebase/firestore/api/query_core.mm]: It's certain because of the underlying knowledge that `query().path` will be even-length if-and-only-if it's a collection group query (because collection group queries are rooted on a document while collection queries are rooted on a collection).  So when we hit this error because `IsDocumentKey(path)` is not true (i.e. `path` is odd-length), there are two possibilities:
1. It's a collection query (odd-length) combined with an even-length document_key.
2. It's a collection group query (even-length) combined with an odd-length document_key.

Option 1 is covered by the prior if statement, since to have an even-length document_key it must contain slashes.  So if we have hit this error it's because of option 2 and that implies it's a collection group query.

If this is too non-obvious, feel free to log a bug against me and I'll go ahead and comment / rework to make it clearer.
